promql/engine.go

```diff
@@ -1499,3 +1518,25 @@ func documentedType(t ValueType) string {
 		return string(t)
 	}
 }
+
+// countingIterator implements the storage.SeriesIterator interface while also
+// counting the number of samples looked up.
+type countingIterator struct {
+	counter *int64
+	it      storage.SeriesIterator
+}
+
+func newCountingIterator(c *int64, it storage.SeriesIterator) *countingIterator {
+	return &countingIterator{c, it}
+}
+
+func (ci countingIterator) Seek(t int64) bool { return ci.it.Seek(t) }
+
+func (ci countingIterator) At() (int64, float64) {
+	atomic.AddInt64(ci.counter, 1)
```

Comment:
Have you checked the overall performance impact? Atomic increments are not equivalent to regular increments. Neither is another layer of interface.

Chances are it's totally fine, but verifying would be important in this case I think.